### PR TITLE
Add placeholder in plugin json

### PR DESCRIPTION
**Pull request by NilsJacobsen**
I added placeholder support for the plugin. 
placeholders parsed with {name} syntax for now.

### CODE REVIEW

**Comment by NilsJacobsen** [comment]
@samuelstroschein something like this?

File: src/plugin.ts [hunk 1]
```diff
@@ -135,14 +137,12 @@ function parseResource(
  * @example
  *  parseMessage("test", "Hello world")
  */
-function parseMessage(id: string, value: string): ast.Message {
-
-  const regex = /({[^}]+})/g;
-  const splitArray = value.split(regex);
+function parseMessage(id: string, value: string, regex?: string): ast.Message {
 
+  const splitArray = value.split(regex || '(?!.*)');
```
Comment (samuelstroschein) [inline src/plugin.ts:142]: why don't you check if regex is undefined and if regex is undefined don't apply placeholder logic?

**Comment by samuelstroschein** [comment]
@NilsJacobsen yeah something like that. I added one remark. Seems like you should apply conditional logic if the placeholder setting has been defined or not instead of a fallback

**Comment by samuelstroschein** [comment]
@NilsJacobsen does the roundtrip test work? 

You seemed to have only updated the `parseMessage` function but not `serializeMessage`

**Comment by NilsJacobsen** [comment]
Roundtrip passed for with and without regex👌🏻

File: src/plugin.ts [hunk 2]
```diff
@@ -173,5 +201,14 @@ function serializeResource(resource: ast.Resource): string {
...
 function serializeMessage(message: ast.Message): [id: string, value: string] {
-  return [message.id.name, (message.pattern.elements[0] as ast.Text).value];
-}
+  const newStringArr = [];
+  for( const element of message.pattern.elements) {
+    if (element.type === "Text") {
+      newStringArr.push(element.value);
+    } else if (element.type === "Placeholder") {
+      newStringArr.push(`{${element.body.name}}`);
+    }
```
Comment (samuelstroschein) [inline src/plugin.ts:210]: The plugin will crash on serialization if the variableReferencePattern is not `{}`. The pattern is hardcoded here. 

1. I recommend adding a test with another placeholder pattern like `[]`. 
2. Update this code to be non-hardcoded. 

2 will be difficult :D

**Review by samuelstroschein — CHANGES_REQUESTED** [review]

File: src/plugin.ts (hunk 2 above)
Comment (NilsJacobsen) [reply src/plugin.ts:210]: Look at that when I return from the Frühlingsfest

Comment (NilsJacobsen) [reply src/plugin.ts]: I used a toupe instead of a regex statement. Because it was hard to parse it back from a regex. So we should now be able to use most of the syntax that they use. What do you think?

Comment (samuelstroschein) [reply src/plugin.ts]: This will eventually break apart as well if more complex variable matches are required. FYI Regex can't be used for parsing and serialization reliably.

But, how about the following: 

```ts
type PluginSettings = {
 variableReferencePattern: {
    parse: Regex,
    serialize: (placeholder: VariableReference) => string 
  }
}

jsonPlugin({
  variableReferencePattern: {
    parse: /{}/,
    serialize: (placeholder) => `{${placeholder.name}}`
  }
})
```

Let people match with regex (because easy) but expose a callback to serialize a variable reference.

Comment (NilsJacobsen) [reply src/plugin.ts]: I have the feeling that this looks pretty difficult to configure. Let's say they need to adapt it, then they have to understand the regex and the implementation to come up with something that passes a roundtrip🙈 in the end we do not now about a lot projects that use something different then {} right so why supporting every edge case from the beginning?

Comment (NilsJacobsen) [reply src/plugin.ts]: Ah wait probably it's should be compatible in the future right. Okey hmmm so if we go with a solution like that, we need to provide an error handling that lets you know what went back if the parse and the serialize does not match.

**Comment by NilsJacobsen** [comment]
So I worked the whole morning on this and really tried to compare the two approaches.
1. regex passing
 ```typescript
 variableReferencePattern: {
    parse: Regex,
    serialize: (placeholder: ast.VariableReference) => string 
  }
  ```
  
  2. pattern with start and end string
  ```typescript
  variableReferencePattern: ["{", "}"]
  ```
  
In the morning I first thought that your arguments convinced me, but during the day it changed again. 
  
**What is bad about the regex passing:**
  - As a enduser u need to define to things that are mentally the same
  - It is not easy to set it up. Regex is a lot harder to enter then just two strings. At least I have problems understanding every regex statement. You can use AI to generate it but still I'm never 100% sure :D I feel like users gonna have the same problem.
  - Users need to provide a parse and serialize input that need to match. When it does not match there will be a roundtrip error. 
  - The config gets bigger
  - Feels more complex
  - It's harder to maintain for us because the complexity gets higher
  - It's pretty hard to setup initially. Like I said I worked the whole morning on it but it was not a nice experience :D I feel like we can not set it up reliable. Had problems with removing the syntax before putting it in the ast. It's the same reason that you already mentioned:
  >Regex can't be used for parsing and serialization reliably.
  
**What could be good about the regex passing:**
  - We could work with variables that have the pattern somewhere in the middle of it. Which is very likely.
  
So I think no it is pretty clean and it seems to work for all of the repos that I have seen. 
What I tested:
{user name}, {{user name}}, [user name], %user name%, %{user name}, %user_name, ::user_name

So I would stick to that approach, cause it's nice and easy. Additionally it gives us freedom of parsing it from one to the other format. What do you think?

**Comment by samuelstroschein** [comment]
@NilsJacobsen Good arguments, approved! 

1. Please update the readme to include documentation about the variableReferencePattern
2. Merge this PR (you should now have the rights to do so)
3. Release version 2.1.0

**Review by samuelstroschein — APPROVED** [review]

**Comment by felixhaeberle** [comment]
@NilsJacobsen @samuelstroschein 
a few question.

did you think about nested placeholders? 

```js
const name = 'John';
console.log(`Hello, ${`my name is ${name}`}.`);
```

and about tag functions?

```js
function myTag(strings, ...values) {
  console.log(strings); // Output: [ 'The product is ', ' USD.', '' ]
  console.log(values); // Output: [ 50 ]
  return strings[0] + values[0] * 0.9 + strings[1];
}
const price = 50;
console.log(myTag`The product is ${price} USD.`);
```

and about js code within the placeholder? especially because of special characters like our dot issue?

```js
const items = ['apple', 'banana', 'cherry'];
console.log(`You have ${items.length} items: ${items.join(', ')}.`);
```

**Comment by NilsJacobsen** [comment]
No,
No,
Yes but not in scope for now.

These things should be realized in specific plugins not maintained by us.

Thanks for bringing this up:)

**Comment by felixhaeberle** [comment]
@NilsJacobsen I think it's pretty common to have nested placeholders / variables.

Have you checked that those syntax doesn't throw bugs?

**Comment by NilsJacobsen** [comment]
I't shouldn't through an error but it will not detect it correctly. But if you want that just fork the repo:) should be an easy change for them but to make it super universal is complex and time consuming

**Comment by felixhaeberle** [comment]
@NilsJacobsen It should at least throw a bug when not detected correctly. Can you fix that?

**Comment by samuelstroschein** [comment]
@felixhaeberle 

The examples you gave are unlocalized string interpolations and not messages with placeholders. 

Your request is well-intentioned but irrelevant to localization/this plugin. Here is an example:

> `console.log(`Hello, ${`my name is ${name}`}.`);`

```ts
// the resulting message is 
"Hello, my name is Samuel."

// a message with a placeholder would be 
"Hello, my name is {name}"

// the consumption of such a message with a placeholder would be 
i("id", { name: "Samuel" })

// and not 
"Hello, ${`my name is ${name}`}." // <- this is an unlocalized string
```

**Comment by felixhaeberle** [comment]
@samuelstroschein I don't think so and I just wanted to make sure that this doesn't break the behavior. It's the same unlocalized/ localized. 

The main question: Do we have good error handling for not supported placeholder structures or does it break the plugins behavior (like the file parsing when some files aren't well formatted translation json files)?

**Comment by samuelstroschein** [comment]
> I don't think so and I just wanted to make sure that this doesn't break the behavior. It's the same unlocalized/ localized.

@felixhaeberle I don't know what you are referring to with this sentence. Could you clarify? 

> The main question: Do we have good error handling for not supported placeholder structures or does it break the plugins behavior (like the file parsing when some files aren't well formatted translation json files)? 

If you are referring to your examples, they won't break the plugin as those are string interpolations. 

Not supported placeholder structures are well not supported. If someone needs support, they should open an issue _or use another plugin that suits their use case alltogether._
